drivers/usb/usbuhci_new/roothub.c

```diff
@@ -0,0 +1,471 @@
+    {
+        portStatus.Suspend = 1;
+    }
+    else
+    {
+        portStatus.Suspend = 0;
+    }
+
+    //if (UhciExtension->HcFlavor == UHCI_Piix4) // check will work after supporting HcFlavor in usbport.
+    if (TRUE)
+    {
+        portStatus.OverCurrent = PortControl.Reserved2 & 1;
+        portStatus.PortPower = (~PortControl.Reserved2 & 1);
+        portChange.OverCurrentIndicatorChange = (PortControl.Reserved2 & 2) != 0;
+    }
+    else
+    {
+        portStatus.OverCurrent = 0;
+        portStatus.PortPower = 1;
+        portChange.OverCurrentIndicatorChange = 0;
+    }
+
+    portStatus.HighSpeedDeviceAttached = 0;
+
+    portStatus.Reset = PortControl.PortReset;
+    portStatus.LowSpeedDeviceAttached = PortControl.LowSpeedDevice;
+    portChange.ConnectStatusChange = PortControl.ConnectStatusChange;
+
+    PortBit = 1 << (Port - 1);
+
+    if (UhciExtension->ResetPortMask & PortBit)
+    {
+        portChange.ConnectStatusChange = 0;
+        portChange.PortEnableDisableChange = 0;
+    }
+    else
+    {
+        portChange.PortEnableDisableChange = PortControl.PortEnableDisableChange;
+    }
+
+    if (UhciExtension->SuspendChangePortMask & PortBit)
+        portChange.SuspendChange = 1;
+
+    if (UhciExtension->ResetChangePortMask & PortBit)
+        portChange.ResetChange = 1;
+
+    PortStatus->PortStatus.Usb20PortStatus = portStatus;
+    PortStatus->PortChange.Usb20PortChange = portChange;
+
+    //DPRINT("UhciRHGetPortStatus: PortControl.AsUSHORT[%x] - %X, PortStatus - %X\n",
+    //       Port,
+    //       PortControl.AsUSHORT,
+    //       PortStatus->AsUlong32);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHGetHubStatus(IN PVOID uhciExtension,
+                   IN PUSB_HUB_STATUS_AND_CHANGE HubStatus)
+{
+    //DPRINT("UhciRHGetHubStatus: ...\n");
+    HubStatus->AsUlong32 = 0;
+    return MP_STATUS_SUCCESS;
+}
+
+VOID
+NTAPI
+UhciRHPortResetComplete(IN PVOID uhciExtension,
+                        IN PVOID pPort)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    ULONG ix;
+    PUHCI_HW_REGISTERS BaseRegister;
+    PUSHORT PortControlRegister;
+    UHCI_PORT_STATUS_CONTROL PortControl;
+    USHORT Port;
+
+    DPRINT("UhciRHPortResetComplete: ...\n");
+
+    BaseRegister = UhciExtension->BaseRegister;
+
+    Port = *(PUSHORT)pPort;
+    ASSERT(Port);
+
+    PortControlRegister = &BaseRegister->PortControl[Port - 1].AsUSHORT;
+    PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+    PortControl.ConnectStatusChange = 0;
+    PortControl.PortEnableDisableChange = 0;
+    PortControl.PortReset = 0;
+
+    WRITE_PORT_USHORT(PortControlRegister, PortControl.AsUSHORT);
+
+    while (UhciHardwarePresent(UhciExtension))
+    {
+        PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+        if (PortControl.PortReset == 0)
+            break;
+    }
+
+    for (ix = 0; ix < 10; ++ix)
+    {
+        KeStallExecutionProcessor(50);
```

Comment:
It's generally used for short waits, either because you're at a high IRQL and can't invoke the scheduler, or to avoid the scheduling overhead (your next timeslice may come in several ms, i.e. much much later than 50µs)

Reply:
Callers of KeDelayExecutionThread must be running at IRQL <= APC_LEVEL.
Callers of KeStallExecutionProcessor can be running at any IRQL.

Reply:
Thanks for the info!